File: 常設中のコンテスト/アルゴリズムと数学 演習問題集/009 - Brute Force 2/brute_force_2.go
package main

import (
    "fmt"
)

func dp(nums []int, s int) bool {
    if len(nums) == 0 {
        return false
    }

    if nums[0] == s {
        return true
    } else if nums[0] > s {
        return dp(nums[1:], s)
    } else {
        return dp(nums[1:], s) || dp(nums[1:], s-nums[0])
    }
}

func main()  {
    var n, s int
    fmt.Scanf("%d %d", &n, &s)

    nums := make([]int, n)
    for i := 0; i < n; i++ {
        fmt.Scanf("%d", &nums[i])
    }

    flag := dp(nums, s)
    if flag {
        fmt.Println("Yes")
    } else {
        fmt.Println("No")
    }
}

File: 常設中のコンテスト/アルゴリズムと数学 演習問題集/009 - Brute Force 2/brute_force_2.py
#!/usr/bin/env python3


def dp(nums, s):
    if len(nums) == 0:
        return False

    if nums[0] == s:
        return True
    elif nums[0] > s:
        return dp(nums[1:], s)
    else:
        return dp(nums[1:], s) or dp(nums[1:], s-nums[0])


if __name__ == "__main__":
    n, s = list(map(int, input().split()))
    nums = list(map(int, input().split()))

    flag = dp(nums, s)
    if flag:
        print("Yes")
    else:
        print("No")
